rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,345 @@
+```
+
+Notably, there is nothing that tells the user how to bypass the check
+using the `insecureUpdate` attribute,
+since if there were, users might very often ignore the error even if valid.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Merging PRs can not be done through GitHub.
+- Setting up PGP signatures will be a bit extra work for the committers.
+- Committers will need to maintain the list of authorized committers.
+- Fetching Nixpkgs with Nix will take around [5 times](#speeding-up-fetching-git) the time, depending on your hardware.
+- Currently building a fork of Nixpkgs (for e.g. a PR) will still work fine, since
+  the latest commit will be trusted when adding a new flake input, or referring
+  to one from the commandline, but updating the flake input will fail
+  since the commits are likely not signed by authorized committers.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Mandating signing only tags is an alternative, though not very useful for most users, since:
+  1) Users of Nixpkgs rarely fix their used revision to a tag.
+  2) Not signing every commit prevents you from using the algorithm described in [updating].
+  Somewhat the same thing can be achieved by the "redirection flake" design described
+  in [future], however.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Should `/.authorized-committers.nix` perhaps not support computation like `flake.nix`?
+  If it supports computation, in Nixpkgs it could be derived from the maintainer list partially.
+- Should we require that repositories that use this mechanism contain
+  the public keys corresponding to the fingerprints to remove the need for key servers?
+
+# Future work
+[future]: #future-work
+
+- Currently, Git still uses SHA-1, which means that you can't
+  completely trust signatures on SHA-1 repositories, since it is a combination
+  of the commit message, parent hashes, and tree hashes, that are seemingly signed.
+  To avoid this, all Git repositories must eventually be transitioned to
+  SHA-256 when Git support for SHA-256 is stabilized. Users can already
+  try SHA-256 if they want extra security for their own repositories.
+  See:
+  - https://git-scm.com/docs/hash-function-transition/
+  - [Git 2.29.0 release notes](https://github.com/git/git/blob/master/Documentation/RelNotes/2.29.0.txt)
+  - [SHA-1 is a Shambles](https://eprint.iacr.org/2020/014.pdf)
+  - https://sha-mbles.github.io/ for a more readable version.
+- The `github` flake input type could potentially be removed since it doesn't seem to be necessary anymore.
+- Instead of an ad-hoc workflow for signing merges, functionality
+  could be added to https://github.com/desktop/desktop and https://github.com/cli/cli
+  to support signing merge commits. See https://github.com/cli/cli/issues/1318.
+- To reduce the amount of time fetching Nixpkgs takes, redirection flakes could be introduced.
```

Comment:
How is this different from just trusting the latest commit on a repo?
How does this compare to just trusting the GitHub signing key instead? It is also a trust proxy. Just like the redirection flake.

Reply:
It isn't for subtle reasons. Let's say i'm managing this redirection repository. I manually change the link to the tarball and the expected hash, and I **check** (through an ad-hoc system) that 1) the new revision is trusted using the usual system and 2) that the contents of the tarball with the specified hash actually match what you'd get from checking out that commit. Then users of my repository would trust that I have done this work.